### PR TITLE
Format multiple selections

### LINKED ISSUE #103745: Cannot format multiple selections
- VSCode Version: 1.47.3
- OS Version: Windows_NT x64 10.0.18363

Steps to Reproduce:

1. use `ctrl+d` select multiple
2. run `format selection` to format them

Demo: https://webmshare.com/play/eadOq

Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by aicest**
This PR fixes #103745

### CODE REVIEW

File: src/vs/editor/contrib/format/formatActions.ts [hunk 1]
```diff
@@ -263,11 +263,13 @@ class FormatSelectionAction extends EditorAction {
...
-			range = new Range(range.startLineNumber, 1, range.startLineNumber, model.getLineMaxColumn(range.startLineNumber));
-		}
-		await instaService.invokeFunction(formatDocumentRangeWithSelectedProvider, editor, range, FormattingMode.Explicit, CancellationToken.None);
+		await editor.getSelections().reduce(async (prev, range: Range) => {
+			await prev;
+			if (range.isEmpty()) {
+				range = new Range(range.startLineNumber, 1, range.startLineNumber, model.getLineMaxColumn(range.startLineNumber));
+			}
+			await instaService.invokeFunction(formatDocumentRangeWithSelectedProvider, editor, range, FormattingMode.Explicit, CancellationToken.None);
+		}, Promise.resolve());
```
Comment (aicest) [inline src/vs/editor/contrib/format/formatActions.ts:272]: Formatting one will fire `onDidChangeModelContent` event, and another will be cancelled.

**Comment by jrieken** [comment]
@aicest Can you merge master and tackle conflicts

**Comment by aicest** [comment]
@jrieken done.

**Review by jrieken — COMMENTED** [review]
Looking good so far but I wonder about the following

* should this be behind a setting - I think no
* should result in a single undo operation instead one per selection - I think yes

For the latter check out the `formatDocumentRangeWithProvider` function. This isn't trivial because the edit operation can happen via the editor or model (when having no editor...) The idea would be to not create undo stop when calling format selection (add argument to control this) but to create undo stops before and after making the edit

File: src/vs/editor/contrib/format/formatActions.ts [hunk 2]
```diff
@@ -267,16 +267,16 @@ class FormatSelectionAction extends EditorAction {
...
 		const model = editor.getModel();
-		let range: Range = editor.getSelection();
-		if (range.isEmpty()) {
-			range = new Range(range.startLineNumber, 1, range.startLineNumber, model.getLineMaxColumn(range.startLineNumber));
-		}
+		const promise = editor.getSelections().reduce(async (prev, range: Range) => {
+			await prev;
+			if (range.isEmpty()) {
+				range = new Range(range.startLineNumber, 1, range.startLineNumber, model.getLineMaxColumn(range.startLineNumber));
+			}
```
Comment (jrieken) [inline src/vs/editor/contrib/format/formatActions.ts:274]: With multiple empty selections on a line this would format a line N times

**Comment by jrieken** [comment]
fyi - I have pushed 379bfe6 which should make implementing this feature much simpler. it will require that you rework this PR tho. let me know if you are still interested

**Pull request closed without merging**